metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py

```diff
@@ -0,0 +1,814 @@
+
+# Logger instance
+LOGGER = logging.getLogger(__name__)
+
+
+class Mapper:
+    """
+    Transform PowerBi concepts Dashboard, Dataset and Tile to DataHub concepts Dashboard, Dataset and Chart
+    """
+
+    class EquableMetadataWorkUnit(MetadataWorkUnit):
+        """
+        We can add EquableMetadataWorkUnit to set.
+        This will avoid passing same MetadataWorkUnit to DataHub Ingestion framework.
+        """
+
+        def __eq__(self, instance):
+            return self.id == instance.id
+
+        def __hash__(self):
+            return id(self.id)
+
+    def __init__(
+        self,
+        config: PowerBiDashboardSourceConfig,
+        reporter: PowerBiDashboardSourceReport,
+    ):
+        self.__config = config
+        self.__reporter = reporter
+
+    @staticmethod
+    def urn_to_lowercase(value: str, flag: bool) -> str:
+        if flag is True:
+            return value.lower()
+
+        return value
+
+    def lineage_urn_to_lowercase(self, value):
+        return Mapper.urn_to_lowercase(
+            value, self.__config.convert_lineage_urns_to_lowercase
+        )
+
+    def assets_urn_to_lowercase(self, value):
+        return Mapper.urn_to_lowercase(value, self.__config.convert_urns_to_lowercase)
+
+    def new_mcp(
+        self,
+        entity_type,
+        entity_urn,
+        aspect_name,
+        aspect,
+        change_type=ChangeTypeClass.UPSERT,
+    ):
+        """
+        Create MCP
+        """
+        return MetadataChangeProposalWrapper(
+            entityType=entity_type,
+            changeType=change_type,
+            entityUrn=entity_urn,
+            aspectName=aspect_name,
+            aspect=aspect,
+        )
+
+    def __to_work_unit(
+        self, mcp: MetadataChangeProposalWrapper
+    ) -> EquableMetadataWorkUnit:
+        return Mapper.EquableMetadataWorkUnit(
+            id="{PLATFORM}-{ENTITY_URN}-{ASPECT_NAME}".format(
+                PLATFORM=self.__config.platform_name,
+                ENTITY_URN=mcp.entityUrn,
+                ASPECT_NAME=mcp.aspectName,
+            ),
+            mcp=mcp,
+        )
+
+    def __to_datahub_dataset(
+        self, dataset: Optional[PowerBiAPI.PowerBIDataset]
+    ) -> List[MetadataChangeProposalWrapper]:
+        """
+        Map PowerBi dataset to datahub dataset. Here we are mapping each table of PowerBi Dataset to Datahub dataset.
+        In PowerBi Tile would be having single dataset, However corresponding Datahub's chart might have many input sources.
+        """
+
+        dataset_mcps: List[MetadataChangeProposalWrapper] = []
+        if dataset is None:
+            return dataset_mcps
+
+        LOGGER.info(
+            f"Converting dataset={dataset.name}(id={dataset.id}) to datahub dataset"
+        )
+
+        for table in dataset.tables:
+            # Create a URN for dataset
+            ds_urn = builder.make_dataset_urn(
+                platform=self.__config.platform_name,
```

Comment:
Where is the platform instance? Is that supported in this source?

Reply:
platform instance is not supported in this source